cmd/frontend/internal/httpapi/graphql.go

```diff
@@ -53,18 +54,42 @@ func actorTypeLabel(isInternal, anonymous bool, requestSource trace.SourceType)
 	return "unknown"
 }
 
-func exceedsLimit(costValue, limitValue int) bool {
-	return costValue > limitValue
+type violationInfo struct {
+	violationType string
+	actual        int
+	limit         int
 }
 
-func writeViolationError(w http.ResponseWriter) error {
+func exceedsLimit(costValue, limitValue int, violationType string) (bool, *violationInfo) {
+	if costValue > limitValue {
+		return true, &violationInfo{
+			violationType: violationType,
+			actual:        costValue,
+			limit:         limitValue,
+		}
+	}
+
+	return false, nil
+}
+
+func writeViolationError(w http.ResponseWriter, info []violationInfo) error {
+	errors := make([]*gqlerrors.QueryError, 0, len(info))
+	for _, info := range info {
+		errors = append(errors, &gqlerrors.QueryError{
+			Message: fmt.Sprintf("Query exceeds maximum %s limit", info.violationType),
+			Extensions: map[string]interface{}{
+				"code":     "errQueryComplexityLimitExceeded",
```

Comment:
Looks like we usually capitalize those in the codebase

```suggestion
				"code":     "ErrQueryComplexityLimitExceeded",
```